gpbft/chain.go

```diff
@@ -176,112 +177,144 @@ func tipSetKeyFromCids(cids []cid.Cid) (TipSetKey, error) {
+// or is nil.
+func (c *ECChain) IsZero() bool {
+	return c == nil || len(c.TipSets) == 0
 }
 
-func (c ECChain) HasSuffix() bool {
-	return len(c) > 1
+func (c *ECChain) HasSuffix() bool {
+	return c.Len() > 1
 }
 
 // Returns the base tipset, nil if the chain is zero.
-func (c ECChain) Base() *TipSet {
+func (c *ECChain) Base() *TipSet {
 	if c.IsZero() {
 		return nil
 	}
-	return &c[0]
+	return c.TipSets[0]
 }
 
 // Returns the suffix of the chain after the base.
 //
 // Returns nil if the chain is zero or base.
-func (c ECChain) Suffix() []TipSet {
+func (c *ECChain) Suffix() []*TipSet {
 	if c.IsZero() {
 		return nil
 	}
-	return c[1:]
+	return c.TipSets[1:]
 }
 
 // Returns the last tipset in the chain.
 // This could be the base tipset if there is no suffix.
 //
 // Returns nil if the chain is zero.
-func (c ECChain) Head() *TipSet {
+func (c *ECChain) Head() *TipSet {
 	if c.IsZero() {
 		return nil
 	}
-	return &c[len(c)-1]
+	return c.TipSets[c.Len()-1]
 }
 
 // Returns a new chain with the same base and no suffix.
 //
 // Returns nil if the chain is zero.
-func (c ECChain) BaseChain() ECChain {
+func (c *ECChain) BaseChain() *ECChain {
 	if c.IsZero() {
 		return nil
 	}
-	return ECChain{c[0]}
+	return &ECChain{
+		TipSets: []*TipSet{c.TipSets[0]},
+	}
 }
 
 // Extend the chain with the given tipsets, returning the new chain.
 //
 // Panics if the chain is zero.
-func (c ECChain) Extend(tips ...TipSetKey) ECChain {
-	// truncate capacity so appending to this chain won't modify the shared slice.
-	c = c[:len(c):len(c)]
+func (c *ECChain) Extend(tips ...TipSetKey) *ECChain {
 	offset := c.Head().Epoch + 1
 	pt := c.Head().PowerTable
+	var extended ECChain
+	extended.TipSets = make([]*TipSet, 0, c.Len()+len(tips))
+	extended.TipSets = append(extended.TipSets, c.TipSets...)
 	for i, tip := range tips {
-		c = append(c, TipSet{
+		extended.TipSets = append(extended.TipSets, &TipSet{
 			Epoch:      offset + int64(i),
 			Key:        tip,
 			PowerTable: pt,
 		})
 	}
-	return c
+	return &extended
 }
 
 // Returns a chain with suffix (after the base) truncated to a maximum length.
 // Prefix(0) returns the base chain.
 //
 // Returns the zero chain if the chain is zero.
-func (c ECChain) Prefix(to int) ECChain {
+func (c *ECChain) Prefix(to int) *ECChain {
 	if c.IsZero() {
 		return nil
 	}
-	length := min(to+1, len(c))
+	length := min(to+1, c.Len())
+	var prefix ECChain
 	// truncate capacity so appending to this chain won't modify the shared slice.
-	return c[:length:length]
+	prefix.TipSets = c.TipSets[:length:length]
+	return &prefix
 }
 
-// Compares two ECChains for equality.
-func (c ECChain) Eq(other ECChain) bool {
-	if len(c) != len(other) {
+// Eq Compares two ECChains for equality.
+// Note that two zero chains are considered equal. See IsZero.
+func (c *ECChain) Eq(other *ECChain) bool {
+	if oneZero, otherZero := c.IsZero(), other.IsZero(); oneZero || otherZero {
+		return oneZero == otherZero
+	}
```

Comment:
another way to write it:
if c.IsZero() && c.IsZero() == other.IsZero() {
     return true
}
but IDK which one is cleaner

Reply:
I think the problem with writing it as `if c.IsZero() && c.IsZero() == other.IsZero() {` is that if `c` is not zero but `other` is zero the equality check progresses to the next set of checks. We don't want that because, "zero" also could mean a nil `ECChain` and that could cause a panic. Right?

Reply:
aah, you are right